lessons/specifics/debugging.md

```diff
@@ -0,0 +1,190 @@
+...
+```
+
+Communicate is clear. Return type of our function `sum_times/1` is different than declared. That because `Enum.sum` return `number` not `integer` and effective return type of `sum_times/1` is `number`. Because `number` is not `integer` so we have error. How to fix it? We could add `round` function that change `number` to `integer`:
+
+```elixir
+@spec sum_times(integer) :: integer
+def sum_times(a) do
+    [1, 2, 3]
+    |> Enum.map(fn el -> el * a end)
+    |> Enum.sum
+    |> round
+end
+```
+
+Finally:
+
+```shell
+$ mix dialyzer
+...
+  Proceeding with analysis... done in 0m0.95s
+done (passed successfully)
+```
+
+Using specifications with tools to static code analysis helps us to make code that is self tested and contains less bugs.  
+
+## Debugging
+
+Sometimes static analysis of code is not enough. We need to know what is a real flow of code, because we would like find bugs. Simplest way is to put in code `IO.puts` to track values and code flow, but this technique is pretty primitive and limited. In Elixir we have could use erlang debugger to debug code. 
+
+Let’s look at a basic module:
+
+```elixir
+defmodule Example do
+
+  def cpu_burns(a, b, c) do
+    x = a * 2
+    y = b * 3
+    z = c * 5
+    
+    x + y + z
+  end
+
+end
+```
+
+Then run `iex`:
+ 
+```bash
+$ iex -S mix
+```
+
+And run debugger:
+
+```elixir
+iex > :debugger.start()
+{:ok, #PID<0.307.0>}
+```
+
+The `:debugger` is Erlang module that give us access to debugger. We could use function `start/1` to configure debugger:
+ 
++ If an argument is path to file, string, then configuration will be read from that file. 
++ If an argument is `:local` or `:global` then debugger will:
+    + `:global` – debugger will interprets code on all known nodes. This is default value.
+    + `:local` – debugger will interprets code only on current node
+
+Next step is to attach our module to debugger:
+
+```elixir
+iex > :int.ni(Example)
+{:module, Example}
+```
+
+The `:int` is an interpreter that gives us possibility to create breakpoint and stepwise execution of code. 
+
+When you start debugger you will see new window like this:
+
+![Debugger Screenshot 1]({{ site.url }}/assets/debugger_1.png)
+
+After attaching our module to debugger it is available in menu on left side:
+
+![Debugger Screenshot 2]({{ site.url }}/assets/debugger_2.png)
+
+### Creating breakpoints
+
+We have two ways to create breakpoint, place where debugger will suspend code execution
+
++ we could call `:int.break/2`,
++ we could use debugger window.
+
+Let's try to create breakpoint from iex:
+
+```elixir
+iex > :int.break(Example, 8)
+:ok
+```
+
+We set breakpoint in line 8 of module `Example` (it is `x + y + z` line). And now when we call our function:
+ 
+```elixir
+iex > Example.cpu_burns(1, 1, 1)
+```
+
+The iex suspends and no result appeared, but in debugger window we see:

```

Comment:
The iex suspends and no result appeared, but in debugger window we see:

-->

Our IEx session suspends and no result appears, but in the debugger we see: